cq-pytest test_secondary_index.py::test_filter_and_limit[False] fails with older Cassandra drivers

@gleb-cloudius and @kostja report that `./test.py --mode dev test_secondary_index` consistently fails on **certain installations** with older versions of the Cassandra driver - it fails on `test_secondary_index.py::test_filter_and_limit[False]`, reporting:

```
            # to fetch the 3 results in tiny one-result pages instead of one page.
            s = cql.prepare(f'SELECT pk FROM {table} WHERE x = 1 AND y = 0 LIMIT 3 ALLOW FILTERING')
            s.fetch_size = 1
>           assert sorted(cql.execute(s)) == [(0,), (2,), (4,)]
E           assert [] == [(0,), (2,), (4,)]
E             Right contains 3 more items, first extra item: (0,)
E             Use -v to get the full diff
test/cql-pytest/test_secondary_index.py:399: AssertionError
```

In this test we do filtering (the "False" parameter says without an index) and retrieve the result in one-item pages. Scylla may, and apparently does, in this case return some _empty_ pages, which exposes a known bug in the Cassandra driver which was fixed last year in https://github.com/datastax/python-driver/commit/1d9077d3f4c937929acc14f45c7693e76dde39a9 and reached the Datastax driver version 3.25.1.

We had exactly the same problem in issue #8203, and the test for it - which requires the fix driver - see `test_filtering.py::test_filtering_contiguous_nonmatching_partition_range` - skips the test if the driver is buggy:

```python
    # Verify that the Python driver is recent enough to contain the fix for
    # the driver bug that caused this issue. If it's too old, this test cannot
    # succeed, and we just skip it. Scylla drivers 3.24.5 or newer contain
    # this fix, Datastax drivers 3.25.1 or newer contain it. The fix was
    # introduced in the following commits:
    # https://github.com/scylladb/python-driver/commit/6ed53d9f7004177e18d9f2ea000a7d159ff9278e,
    # https://github.com/datastax/python-driver/commit/1d9077d3f4c937929acc14f45c7693e76dde39a9
    scylla_driver = 'Scylla' in DRIVER_NAME
    driver_version = tuple(int(x) for x in DRIVER_VERSION.split('.'))
    if (scylla_driver and driver_version < (3, 24, 5) or
            not scylla_driver and driver_version <= (3, 25, 0)):
        pytest.skip("Python driver too old to run this test")
```

We should do the same in the test which is failing now.

--- a/test/cql-pytest/conftest.py
+++ b/test/cql-pytest/conftest.py
@@ -14,6 +14,7 @@ import pytest
 from cassandra.auth import PlainTextAuthProvider
 from cassandra.cluster import Cluster, ConsistencyLevel, ExecutionProfile, EXEC_PROFILE_DEFAULT
 from cassandra.policies import RoundRobinPolicy
+from cassandra.connection import DRIVER_NAME, DRIVER_VERSION
 import ssl
 
 from util import unique_name, new_test_table
@@ -141,6 +142,24 @@ def check_pre_raft(cql):
 def fails_without_raft(request, check_pre_raft):
     if check_pre_raft:
         request.node.add_marker(pytest.mark.xfail(reason='Test expected to fail without Raft experimental feature on'))
+
+# Older versions of the Cassandra driver had a bug where if Scylla returns
+# an empty page, the driver would immediately stop reading even if this was
+# not the last page. Some tests which filter out most of the results can end
+# up with some empty pages, and break on buggy versions of the driver. These
+# tests should be skipped when using a buggy version of the driver. This is
+# the purpose of the following fixture.
+# This driver bug was fixed in Scylla driver 3.24.5 and Datastax driver
+# 3.25.1, in the following commits:
+# https://github.com/scylladb/python-driver/commit/6ed53d9f7004177e18d9f2ea000a7d159ff9278e,
+# https://github.com/datastax/python-driver/commit/1d9077d3f4c937929acc14f45c7693e76dde39a9
+@pytest.fixture(scope="function")
+def driver_bug_1():
+    scylla_driver = 'Scylla' in DRIVER_NAME
+    driver_version = tuple(int(x) for x in DRIVER_VERSION.split('.'))
+    if (scylla_driver and driver_version < (3, 24, 5) or
+            not scylla_driver and driver_version <= (3, 25, 0)):
+        pytest.skip("Python driver too old to run this test")
 
 # TODO: use new_test_table and "yield from" to make shared test_table
 # fixtures with some common schemas.

--- a/test/cql-pytest/test_filtering.py
+++ b/test/cql-pytest/test_filtering.py
@@ -14,7 +14,6 @@ import pytest
 import re
 from util import new_test_table, new_type, user_type
 from cassandra.protocol import InvalidRequest
-from cassandra.connection import DRIVER_NAME, DRIVER_VERSION
 from cassandra.query import UNSET_VALUE
 
 # When filtering for "x > 0" or "x < 0", rows with an unset value for x
@@ -49,19 +48,7 @@ def test_filter_on_unset(cql, test_keyspace):
 # this test.
 
 # Reproducer for issue #8203, partition-range (whole-table) scan case
-def test_filtering_contiguous_nonmatching_partition_range(cql, test_keyspace):
-    # Verify that the Python driver is recent enough to contain the fix for
-    # the driver bug that caused this issue. If it's too old, this test cannot
-    # succeed, and we just skip it. Scylla drivers 3.24.5 or newer contain
-    # this fix, Datastax drivers 3.25.1 or newer contain it. The fix was
-    # introduced in the following commits:
-    # https://github.com/scylladb/python-driver/commit/6ed53d9f7004177e18d9f2ea000a7d159ff9278e,
-    # https://github.com/datastax/python-driver/commit/1d9077d3f4c937929acc14f45c7693e76dde39a9
-    scylla_driver = 'Scylla' in DRIVER_NAME
-    driver_version = tuple(int(x) for x in DRIVER_VERSION.split('.'))
-    if (scylla_driver and driver_version < (3, 24, 5) or
-            not scylla_driver and driver_version <= (3, 25, 0)):
-        pytest.skip("Python driver too old to run this test")
+def test_filtering_contiguous_nonmatching_partition_range(cql, test_keyspace, driver_bug_1):
     # The bug depends on the amount of data being scanned passing some
     # page size limit, so it doesn't matter if the reproducer has a lot of
     # small rows or fewer long rows - and inserting fewer long rows is
@@ -81,12 +68,7 @@ def test_filtering_contiguous_nonmatching_partition_range(cql, test_keyspace):
         assert list(cql.execute(f"SELECT p FROM {table} WHERE v={v} ALLOW FILTERING")) == [(p,)]
 
 # Reproducer for issue #8203, single-partition scan case
-def test_filtering_contiguous_nonmatching_single_partition(cql, test_keyspace):
-    scylla_driver = 'Scylla' in DRIVER_NAME
-    driver_version = tuple(int(x) for x in DRIVER_VERSION.split('.'))
-    if (scylla_driver and driver_version < (3, 24, 5) or
-            not scylla_driver and driver_version <= (3, 25, 0)):
-        pytest.skip("Python driver too old to run this test")
+def test_filtering_contiguous_nonmatching_single_partition(cql, test_keyspace, driver_bug_1):
     count = 100
     long='x'*60000
     with new_test_table(cql, test_keyspace,

--- a/test/cql-pytest/test_secondary_index.py
+++ b/test/cql-pytest/test_secondary_index.py
@@ -373,7 +373,7 @@ def test_index_in_restriction(cql, test_keyspace, cassandra_bug):
 # returned fewer than the requested number of rows.
 @pytest.mark.parametrize("use_index", [
         pytest.param(True, marks=pytest.mark.xfail(reason="#10649")), False])
-def test_filter_and_limit(cql, test_keyspace, use_index):
+def test_filter_and_limit(cql, test_keyspace, use_index, driver_bug_1):
     with new_test_table(cql, test_keyspace, 'pk int primary key, x int, y int') as table:
         if use_index:
             cql.execute(f'CREATE INDEX ON {table}(x)')

